Source/WebKit/UIProcess/mac/WebViewImpl.mm

```diff
@@ -5069,6 +5023,68 @@ static BOOL shouldUseHighlightsForMarkedText(NSAttributedString *string)
     mergedHighlights.shrinkToFit();
     return mergedHighlights;
 }
+#endif
+
+static Vector<WebCore::CompositionUnderline> compositionUnderlines(NSAttributedString *string)
+{
+    if (!string.length)
+        return { };
+
+    Vector<WebCore::CompositionUnderline> mergedUnderlines;
+
+#if HAVE(INLINE_PREDICTIONS) || HAVE(REDESIGNED_TEXT_CURSOR)
+    Vector<WebCore::CompositionUnderline> underlines;
+
+    [string enumerateAttributesInRange:NSMakeRange(0, string.length) options:0 usingBlock:[&underlines](NSDictionary<NSAttributedStringKey, id> *attributes, NSRange range, BOOL *) {
+        NSNumber *style = [attributes objectForKey:NSUnderlineStyleAttributeName];
+        if (!style)
+            return;
+
+        NSColor *underlineColor = attributes[NSUnderlineColorAttributeName];
+        bool isClear = [underlineColor isEqual:NSColor.clearColor];
+
+        if (!isClear)
+            underlines.append({ static_cast<unsigned>(range.location), static_cast<unsigned>(NSMaxRange(range)), WebCore::CompositionUnderlineColor::GivenColor, WebCore::Color::black, style.intValue > 1 });
+    }];
+
+    std::sort(underlines.begin(), underlines.end(), [](auto& a, auto& b) {
```

Comment:
I see similarities between this code and `WKContentViewInteraction.mm`, we should probably refactor at some point (not in this PR).

Reply:
oh ya definitely; that was my plan, I just forgot to mention it 😅